sr_error_reporter/src/UnderactuationErrorReporter.cpp

```diff
@@ -0,0 +1,155 @@
+  trajectory_subsriber_right_ = node_handle.subscribe("/rh_trajectory_controller/command", 1,
+    &UnderactuationErrorReporter::trajectory_callback, this, ros::TransportHints().tcpNoDelay());
+
+  robot_model_loader_.reset(new robot_model_loader::RobotModelLoader("robot_description"));
+  robot_state_.reset(new robot_state::RobotState(robot_model_loader_->getModel()));
+  robot_state_->setToDefaultValues();
+}
+
+ros::Publisher UnderactuationErrorReporter::get_or_create_publisher(std::string link_name)
+{
+  auto iterator = error_publishers_.find(link_name);
+  if (iterator != error_publishers_.end())
+  {
+    return iterator->second;
+  }
+  ros::Publisher publister = node_handle_.advertise<sr_error_reporter::UnderactuationError>(
+    "/underactuation_error/" + link_name, 1);
+  error_publishers_[link_name] = publister;
+  return publister;
+}
+
+void UnderactuationErrorReporter::update_kinematic_model(
+  std::map<std::string, double>& joint_positions,
+  std::map<std::string, geometry_msgs::Transform>& transforms)
+{
+  robot_state_->setToDefaultValues();
+  for (auto& side : sides_)
+  {
+    for (auto& finger : include_fingers_)
+    {
+      auto j1 = joint_positions.find(side + finger.first + "distal");
+      auto j2 = joint_positions.find(side + finger.first + "middle");
+      std::vector<double> positions = {0, 0, j2->second, j1->second};
+      robot_state_->setJointGroupPositions(side + finger.second, positions);
+    }
+  }
+  for (auto& side : sides_)
+  {
+    for (auto& finger : include_fingers_)
+    {
+      std::string link_name = side + finger.first + "tip";
+      // Forward kinemetics
+      const Eigen::Affine3d &end_effector_state = robot_state_->getGlobalLinkTransform(link_name);
+      geometry_msgs::Transform transform;
+      tf::transformEigenToMsg(end_effector_state, transform);
+      transforms[link_name] = transform;
+    }
+  }
+}
+
+void UnderactuationErrorReporter::publish_error()
+{
+  for (auto& actual : actual_tip_transforms_)
+  {
+    std::string link_name = actual.first;
+    auto desired = desired_tip_transforms_.find(link_name);
+    if (desired != desired_tip_transforms_.end())
+    {
+      double x = actual.second.translation.x - desired->second.translation.x;
+      double y = actual.second.translation.y - desired->second.translation.y;
+      double z = actual.second.translation.z - desired->second.translation.z;
+      double error = std::sqrt(x * x + y * y + z * z);
+
+      sr_error_reporter::UnderactuationError underactuation_error;
+      underactuation_error.header.stamp = ros::Time::now();
+      underactuation_error.error = error;
+      get_or_create_publisher(link_name).publish(underactuation_error);
+    }
+  }
+}
+
+void UnderactuationErrorReporter::update_joint_position(
+  std::map<std::string, double>& joint_positions,
+  std::string name,
+  double position)
+{
+  if (name.size() < 7)
```

Comment:
what is 7? maybe add a const?

Reply:
Added a comment in bf3029a16310672b6159f2cfc6a2e764489858e1 with a example of possible name. Hopefully that makes it clear which length is expected.